src/parallel/stderr.rs

```diff
@@ -0,0 +1,74 @@
+/// Helpers functions for [ChildStderr].
+use std::{io::Read, process::ChildStderr};
+
+use crate::Error;
+
+#[cfg(all(not(unix), not(windows)))]
+compile_error!("Only unix and windows support non-blocking pipes! For other OSes, disable the parallel feature.");
+
+#[allow(unused_variables)]
+pub fn set_non_blocking(stderr: &mut ChildStderr) -> Result<(), Error> {
+    // On Unix, switch the pipe to non-blocking mode.
+    // On Windows, we have a different way to be non-blocking.
+    #[cfg(unix)]
+    {
+        use std::os::unix::io::AsRawFd;
+        let fd = stderr.as_raw_fd();
+        debug_assert_eq!(
+            unsafe { libc::fcntl(fd, libc::F_GETFL, 0) },
+            0,
+            "stderr should have no flags set"
+        );
+
+        if unsafe { libc::fcntl(fd, libc::F_SETFL, libc::O_NONBLOCK) } != 0 {
+            return Err(Error::new(
+                crate::ErrorKind::IOError,
+                format!(
+                    "Failed to set flags for child stderr: {}",
+                    std::io::Error::last_os_error()
+                ),
+            ));
+        }
+    }
+
+    Ok(())
+}
+
+pub fn read_non_blocked(stderr: &mut ChildStderr, buf: &mut [u8]) -> Result<usize, std::io::Error> {
+    // On Windows, take a peek at the pipe to see if there's data.
+    #[cfg(windows)]
+    {
+        use crate::windows::windows_sys::PeekNamedPipe;
+        use std::os::windows::io::AsRawHandle;
+        use std::ptr::null_mut;
+        let mut bytes_available = 0;
+        unsafe {
+            if PeekNamedPipe(
+                stderr.as_raw_handle(),
+                null_mut(),
+                0,
+                null_mut(),
+                &mut bytes_available,
+                null_mut(),
+            ) == 0
+            {
+                return Err(std::io::Error::last_os_error());
+            }
+        }
+        if bytes_available == 0 {
```

Comment:
I wonder if it makes sense to pass the `Vec` here, then `.reserve(bytes_available)` to make sure all remaining data is read.

Reply:
With the change to use `FIONREAD`, I've modified the code to split checking available data from doing the actual read, so it's pretty easy to reserve at least the bytes currently available.